maze_gen/mazecetric/maze.py

```diff
@@ -0,0 +1,71 @@
+from maze_gen import Maze
+from maze_gen.mazecetric.cell import Cell, RuleSet, Board
+
+
+class MazecetricMaze(Maze):
+    """
+    FOR SHAIIBON:
+    The main thing you need to implement is the `generate()` and `__str__()` methods. The first - obviously, should
+    generate the previously initialized maze (and change the object's inner state, not return anything). The second will
+    be used for maze printing, so it needs to follow the strict interface - the returned string, representing the
+    maze, needs to include the outside walls and inner passages. The walls should be always represented by '#'
+    character, and the passages by the '.' character. Remember that the total size of the maze (i.e. the number of
+    characters representing the single row in the output string) will always be odd. Good luck!
+    """
+
+    ALIVE_CHANCE = 0.5
+
+    def __init__(self, size: int):
+        super(MazecetricMaze, self).__init__(size, "Mazecetric")
+        self.rule_set = RuleSet("Mazecetric", a_rules=(1, 2, 3, 4), d_rules=(3,))
+        self.board = Board(self.ALIVE_CHANCE, self.total_grid_size)
+
+    def will_survive(self, cell) -> bool:
+        c = self.count_alive_neighbours(cell)
+        if not cell.is_alive and c in self.rule_set.dead_rules:
+            return True
+        elif cell.is_alive and c in self.rule_set.alive_rules:
+            return True
+        return False
+
+    def generate(self):
+        for _ in range(100):
+            next_gen = Board(0, self.total_grid_size)
+            for i, row in enumerate(self.board):
+                for cell in row:
+                    # dont touch the side walls
+                    if (
+                        0 < cell.x < self.total_grid_size - 1
+                        and 0 < cell.y < self.total_grid_size - 1
+                    ):
+                        next_gen.set_cell(cell.x, cell.y, self.will_survive(cell))
+            self.board = next_gen
+
+    def count_alive_neighbours(self, cell):
```

Comment:
I would approach this a bit differently - make a method `get_neighbours(cell: Cell)` that returns a list of neighbours that looks roughly like this:
```python
def get_neighbours(cell: Cell) -> list[Cell]:
    return list(filter(None, [
        self.board,get_cell(x -1, y -1),
        ...
        self.board.get_cell(x + 1, y + 1),
    ]))
```
and make `Board.get_cell` return either a `Cell` (if the indices are right) or `None` (otherwise).

Reply:
it will be more clear and explicit, right now the implementation is a little unintuitive